Inherent static methods can be called with a non-well-formed Self-type

**UPDATE:** This has been fixed, but a test is still needed. [See the comment below for instructions.](https://github.com/rust-lang/rust/issues/28848#issuecomment-294848720)

-------

## STR

``` Rust
struct Foo<'a, 'b: 'a>(&'a &'b ());

impl<'a, 'b> Foo<'a, 'b> {
    fn xmute(a: &'b ()) -> &'a () {
        unreachable!()
    }
}

pub fn foo<'a, 'b>(u: &'b ()) -> &'a () {
    Foo::<'a, 'b>::xmute(u)
}

fn main() {}
```

Here a static method on `Foo` is called with unrelated lifetimes `'a` and `'b`, potentially violating the `'b: 'a` bound. As `Foo::xmute` is not allowed to assume `Foo` is well-formed, this does _not_ cause a soundness issue, but is still a potential back-compat hazard.

cc @nikomatsakis

This no longer compiles with the following error, which I think means that this is fixed. Can someone confirm?

```
error[E0478]: lifetime bound not satisfied
  --> /tmp/t:10:5
   |
10 |     Foo::<'a, 'b>::xmute(u)
   |     ^^^^^^^^^^^^^^^^^^^^
   |
note: lifetime parameter instantiated with the lifetime 'b as defined on the body at 9:40
  --> /tmp/t:9:41
   |
9  |   pub fn foo<'a, 'b>(u: &'b ()) -> &'a () {
   |  _________________________________________^ starting here...
10 | |     Foo::<'a, 'b>::xmute(u)
11 | | }
   | |_^ ...ending here
note: but lifetime parameter must outlive the lifetime 'a as defined on the body at 9:40
  --> /tmp/t:9:41
   |
9  |   pub fn foo<'a, 'b>(u: &'b ()) -> &'a () {
   |  _________________________________________^ starting here...
10 | |     Foo::<'a, 'b>::xmute(u)
11 | | }
   | |_^ ...ending here

error: aborting due to previous error
```

--- a/src/test/compile-fail/issue-28848.rs
+++ b/src/test/compile-fail/issue-28848.rs
@@ -1,0 +1,23 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+struct Foo<'a, 'b: 'a>(&'a &'b ());
+
+impl<'a, 'b> Foo<'a, 'b> {
+    fn xmute(a: &'b ()) -> &'a () {
+        unreachable!()
+    }
+}
+
+pub fn foo<'a, 'b>(u: &'b ()) -> &'a () {
+    Foo::<'a, 'b>::xmute(u) //~ ERROR lifetime bound not satisfied
+}
+
+fn main() {}

